src/datasets/arrow_writer.py

```diff
@@ -549,14 +544,12 @@ def write_batch(
         try_features = self._features if self.pa_writer is None and self.update_features else None
         arrays = []
         inferred_features = Features()
+        # preserve the order the columns
         if self.schema:
             schema_cols = set(self.schema.names)
-            common_cols, extra_cols = [], []
-            for col in batch_examples:
-                if col in schema_cols:
-                    common_cols.append(col)
-                else:
-                    extra_cols.append(col)
+            batch_cols = batch_examples.keys()  # .keys() preserves the order (unlike set)
+            common_cols = [col for col in self.schema.names if col in batch_cols]
+            extra_cols = [col for col in batch_cols if col not in schema_cols]
             cols = common_cols + extra_cols
         else:
             cols = list(batch_examples)
```

Comment:
I feel like we should really avoid this extra copy, especially if the inner iterable is large.

Reply:
This negligible optimization (`write_examples_on_file` and `write_batch` are called every `writer_batch_size`) goes against good code practices.

We wouldn't use Python for this project if we wanted to optimize every aspect of the API.

Reply:
I'm not sure it's negligible. #6636's OP stated:

> I work with bioinformatics data and often these tables have thousands and even tens of thousands of features. 

We'd make a copy of tens of thousands of strings for every batch, for every processing step (e.g., a `map`).

And it's easy to remove (just `cols = batch_samples`, instead of copying it into a list).

Among other things, this library is about large data processing efficiency, so I think it'd be nice to consider it.